### PR TITLE
azure-eventhub: expand input sanitizer's invalid JSON coverage

### PR DESCRIPTION
**Pull request by zmoog**
## Proposed commit message

- Rework the input's invalid JSON sanitization feature to support configurable sanitizers
- Migrate existing sanitizers to the new format
- Add a new `replace_all` sanitizer
- Add a new `invalid_json_messages_total` input metric to count the number of messages that contain invalid JSON strings.

The `replace_all` sanitizer replaces all the occurrences of substring matching the regex expression `pattern` with a fixed literal string `replacement`.

Here is a sample configuration:

```yaml
sanitizers:
- type: replace_all
  spec:
    pattern: '\[\s*([^\[\]{},\s]+(?:\s+[^\[\]{},\s]+)*)\s*\]'
    replacement: "{}"
```

For example, if the diagnostic settings send the following message:

```json
{
    "AppImage": "orcas/postgres_standalone_16_u18:38.1.240825",
    "AppType": "PostgreSQL",
    "properties": [
        218 B blob data
    ]
}
```

With the previous sample configuration, the input will replace the invalid JSON, updating the message to the following version:

```json
{
    "AppImage": "orcas/postgres_standalone_16_u18:38.1.240825",
    "AppType": "PostgreSQL",
    "properties": {}
}
```

The `replace_all` sanitizer aims to restore JSON syntax validity by replacing invalid, unfixable JSON with literal values that make sense for the context.

The invalid JSON sanitization became a thing when we faced Azure services producing malformed JSON documents.

Here is a list of previous occurrences of this problem:

- https://github.com/elastic/beats/issues/34092
- https://github.com/elastic/azuremarketplacedev/issues/190

In all cases, we first reached out to Azure support, but they didn't fix the invalid JSON.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Closes https://github.com/elastic/beats/issues/40641

### CODE REVIEW

**Comment by zmoog** [comment]
Here is a possible config setting for the new Sanitizer API:

```yaml
- module: azure
  # All logs
  activitylogs:
    enabled: true
    var:
      # eventhub name containing the activity logs, overwrite he default value if the logs are exported in a different eventhub
      eventhub: "whatever"
      # consumer group name that has access to the event hub, we advise creating a dedicated consumer group for the azure module
      consumer_group: "$Default"
      # the connection string required to communicate with Event Hubs, steps to generate one here https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-get-connection-string
      connection_string: "Endpoint=sb://""
      # the name of the storage account the state/offsets will be stored and updated
      storage_account: "losqradar"
      # the name of the storage account container you would like to store the offset information in.
      storage_account_container: "filebeat-activitylogs"
      # the storage account key, this key will be used to authorize access to data in your storage account
      storage_account_key: ""

      sanitizers:
        - type: new_lines
        - type: single_quote
        - type: regex
          spec:
            # The regex to use to sanitize the logs
            pattern: '\[\s*([^\[\]{},\s]+(?:\s+[^\[\]{},\s]+)*)\s*\]'
            # The replacement to use for the regex
            replacement: '[${1}]'
```

File: x-pack/filebeat/input/azureeventhub/sanitizers.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	}
+
+	if _, ok := s.spec["replacement"].(string); !ok {
+		return errors.New("replacement format must be a string")
+	}
+
+	s.replacement = s.spec["replacement"].(string)
+
+	return nil
+}
```
Comment (gpop63) [inline x-pack/filebeat/input/azureeventhub/sanitizers.go:201]: Do you think its worth creating a helper function that gets a key from the `spec` map?

```suggestion
func (s *replaceAllSanitizer) Init() error {
    if s.spec == nil {
        return errors.New("missing sanitizer spec")
    }

    pattern, err := getStringFromSpec(s.spec, "pattern")
    if err != nil {
        return fmt.Errorf("invalid pattern: %w", err)
    }

    re, err := regexp.Compile(pattern)
    if err != nil {
        return fmt.Errorf("invalid regex pattern: %w", err)
    }
    s.re = re

    replacement, err := getStringFromSpec(s.spec, "replacement")
    if err != nil {
        return fmt.Errorf("invalid replacement: %w", err)
    }
    s.replacement = replacement

    return nil
}

func getStringFromSpec(spec map[string]interface{}, key string) (string, error) {
    value, ok := spec[key]
    if !ok {
        return "", fmt.Errorf("missing %s", key)
    }

    strValue, ok := value.(string)
    if !ok {
        return "", fmt.Errorf("%s must be a string", key)
    }

    return strValue, nil
}
```

**Review by gpop63 — COMMENTED** [review]
This new API you created looks very nice! Do we need a changelog entry for this?

File: x-pack/filebeat/input/azureeventhub/sanitizers.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+// ----------------------------------------------------------------------------
+// Convenience builder functions
+// ----------------------------------------------------------------------------
+
+// newSanitizer creates a new sanitizer based on the provided spec.
+func newSanitizer(spec SanitizerSpec) (Sanitizer, error) {
+	var s Sanitizer
+
+	switch spec.Type {
+	case "new_lines":
```
[the remark this answers is not in the document]
Comment (zmoog) [reply x-pack/filebeat/input/azureeventhub/sanitizers.go:47]: Yeah, I omitted a final engineering pass on the actual sanitizer implementation, like the `Init()` function on the `replace_all`. 

On them!

File: x-pack/filebeat/input/azureeventhub/config.go [hunk 3]
```diff
@@ -121,7 +149,8 @@ func (conf *azureInputConfig) Validate() error {
 	}
 
 	// log a warning for each sanitization option not supported
-	for _, opt := range conf.SanitizeOptions {
+	for _, opt := range conf.LegacySanitizeOptions {
+		logger.Warnw("legacy sanitization `sanitize_options` options are deprecated and will be removed in future releases; use the `sanitizers` option instead", "option", opt)
```
Comment (kaiyan-sheng) [inline x-pack/filebeat/input/azureeventhub/config.go:153]: Should we also mark in the documentation about the `sanitize_options` config will be removed in the future? Maybe specify in 9.0?

**Review by kaiyan-sheng — APPROVED** [review]

File: x-pack/filebeat/input/azureeventhub/config.go (hunk 3 above)
Comment (Kavindu-Dodan) [reply x-pack/filebeat/input/azureeventhub/config.go:153]: +1 on this suggestion :) we should define the removal version and create a follow-up ticket so we don't forget about this :)

File: x-pack/filebeat/input/azureeventhub/sanitizers.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+// ----------------------------------------------------------------------------
+
+// SanitizerSpec defines a sanitizer configuration.
+//
+// Sanitizers can use the spec field to provide additional
+// configuration.
+type SanitizerSpec struct {
+	Type string                 `config:"type"`
+	Spec map[string]interface{} `config:"spec"`
+}
```
Comment (Kavindu-Dodan) [inline x-pack/filebeat/input/azureeventhub/sanitizers.go:33]: +1 on this. This gives us flexibility to add custom sanitizer implementations :)

**Review by Kavindu-Dodan — APPROVED** [review]

File: x-pack/filebeat/input/azureeventhub/config.go (hunk 3 above)
Comment (zmoog) [reply x-pack/filebeat/input/azureeventhub/config.go:153]: 9.0 is handy for deprecation purposes; I'm creating a note to track this.

**Review by lucian-ioan — APPROVED** [review]
LGTM. Great improvement over the original implementation :fire:

**Pull request merged**